Automattic/woocommerce-payments
# RPP: Added error states

Fixes #7425 

#### Changes proposed in this Pull Request

This PR introduces error states and their management. It ensures that all exceptions, including minor ones, are effectively handled. The core principle is to avoid returning exceptions from states. Instead, states should consistently return their designated state. This approach eliminates the need for exception handling in services utilizing these states or in the gateway, where they were previously handled.

Error states:

- `AbstractPaymentErrorState` : Abstract error class that holds common or reusable logic for all error states
- `PaymentErrorState`: Error state that is used when an error occurs on the buyer's side, which cannot be fixed through code.
- `PaymentRequestErrorState`: Handles `PaymentRequestException` that can occur when context is populated from the request.
- `SystemErrorState`: Equivalent of \Exception we have in PHP but used for our state machine. Look at this state as a default error state when an exception or any unwanted flow occurs, where we don't have a specific error state to transition.
- `WooPaymentsApiServerErrorState`: Error state that is used when server API throws an exception. I used WooPayments prefix to make it more clearer that this is the error state for API server errors on WooCommerce Payments server.











#### Testing instructions





* Make sure that all exceptions are handled by the error states. In other words, if an exception is triggered in state, the state needs to handle it and return error state. The only exception that can be handled without the error state is `StateTransitionException` which can occur outside state processing or within the catch block, where we transition from state to new error state.
* All error states should extend the base or abstract error class.
* Make sure that all error states have their logging and mark order as failed function configured property, and according to the business logic. What that means is that each error state can have disabled logging or disabled marking order as failed.

-------------------


**Post merge**
